cli/admin/call.go

```diff
@@ -0,0 +1,187 @@
+package admin
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/apex/log"
+	"github.com/tarantool/cartridge-cli/cli/project"
+
+	"github.com/spf13/pflag"
+
+	"github.com/tarantool/cartridge-cli/cli/common"
+	"github.com/tarantool/cartridge-cli/cli/templates"
+)
+
+const (
+	argTypeString = "string"
+	argTypeNumber = "number"
+	argTypeBool   = "boolean"
+)
+
+type FuncCallArg struct {
+	Name string
+	Type string
+
+	StringValue string
+	NumberValue int
+	BoolValue   bool
+
+	Changed bool
+}
+
+func adminFuncCall(conn net.Conn, funcName string, flagSet *pflag.FlagSet, args []string) error {
+	funcCallArgs, err := getFuncCallArgs(conn, funcName, flagSet, args)
+	if err != nil {
+		return fmt.Errorf("Failed to parse function call args: %s", err)
+	}
+
+	argsSerialized, err := serializeArgs(funcCallArgs)
+	if err != nil {
+		return fmt.Errorf("Failed to serialize function args: %s", err)
+	}
+
+	callFuncBody, err := templates.GetTemplatedStr(&adminCallFuncBodyTmpl, map[string]string{
+		"AdminCallFuncName": adminCallFuncName,
+		"FuncName":          funcName,
+		"Args":              argsSerialized,
+	})
+
+	callResRaw, err := common.EvalTarantoolConn(conn, callFuncBody)
+	if err != nil {
+		return fmt.Errorf("Failed to call %q: %s", funcName, err)
+	}
+
+	needReturnValueWarn := false
+
+	switch callRes := callResRaw.(type) {
+	case string:
+		log.Info(callRes)
+	case []interface{}:
+		for _, callResLineRaw := range callRes {
+			switch callResLine := callResLineRaw.(type) {
+			case string:
+				log.Info(callResLine)
+			default:
+				needReturnValueWarn = true
+				log.Infof("%v\n", callResLine)
+			}
+		}
+	default:
+		needReturnValueWarn = true
+		log.Infof("%v\n", callRes)
+	}
+
+	if needReturnValueWarn {
+		log.Warnf("Admin function %q should return string or string array value", funcName)
+	}
+
+	return nil
+}
+
+func getFuncCallArgs(conn net.Conn, funcName string, flagSet *pflag.FlagSet, args []string) ([]FuncCallArg, error) {
+	helpResRawMap, err := getFuncHelpRawMap(funcName, conn)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get function %q signature: %s", funcName, err)
```

Comment:
Here I don't understand a root of problem, what should I (as user) do? I suggest to more verbose error message.

Reply:
Added `getCliExtError` to format such messages as a `cartridge-cli-extentions` error.